Notebook baseado no livro Statistics For Machine Learning

In [1]:
import os
os.chdir("C:\\Users\\ViniciusNascimento\\Desktop\\ML\\data\\ml-latest-small")

from IPython.display import display

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Funções para nos auxiliar na exploração dos dados:

In [2]:
def descreve_df(df):
    display(df.sample(n=5, random_state=123))
    display(df.describe())
    display(df.info())

In [3]:
ratings = pd.read_csv('ratings.csv')
descreve_df(ratings)

userId  movieId  rating   timestamp
16705     105    77455     5.0  1446610793
3692       22     4034     4.5  1268726741
74367     474     4912     3.0  1007320701
49894     319   136864     4.5  1461454241
84800     550   115713     4.5  1488728279

userId        movieId         rating     timestamp
count  100836.000000  100836.000000  100836.000000  1.008360e+05
mean      326.127564   19435.295718       3.501557  1.205946e+09
std       182.618491   35530.987199       1.042529  2.162610e+08
min         1.000000       1.000000       0.500000  8.281246e+08
25%       177.000000    1199.000000       3.000000  1.019124e+09
50%       325.000000    2991.000000       3.500000  1.186087e+09
75%       477.000000    8122.000000       4.000000  1.435994e+09
max       610.000000  193609.000000       5.000000  1.537799e+09

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 4 columns):
userId       100836 non-null int64
movieId      100836 non-null int64
rating       100836 non-null float64
timestamp    100836 non-null int64
dtypes: float64(1), int64(3)
memory usage: 3.1 MB


None

In [4]:
movies = pd.read_csv('movies.csv')
descreve_df(movies)

movieId                                title                     genres
264       304                     Roommates (1995)               Comedy|Drama
8903   134881                  Love & Mercy (2014)                      Drama
2713     3639  Man with the Golden Gun, The (1974)  Action|Adventure|Thriller
3817     5347                   Deuces Wild (2002)                      Drama
3024     4043                At Close Range (1986)                Crime|Drama

movieId
count    9742.000000
mean    42200.353623
std     52160.494854
min         1.000000
25%      3248.250000
50%      7300.000000
75%     76232.000000
max    193609.000000

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 3 columns):
movieId    9742 non-null int64
title      9742 non-null object
genres     9742 non-null object
dtypes: int64(1), object(2)
memory usage: 228.4+ KB


None

Vamos dividir os gêneros em Dummies para facilitar o processamento futuro

In [5]:
dummies = movies['genres'].str.get_dummies(sep='|')
movies.drop(columns=['genres'], inplace=True)
#print(dummies.head())

movies = pd.concat([movies, dummies], axis=1)
descreve_df(movies)

movieId                                title  (no genres listed)  \
264       304                     Roommates (1995)                   0   
8903   134881                  Love & Mercy (2014)                   0   
2713     3639  Man with the Golden Gun, The (1974)                   0   
3817     5347                   Deuces Wild (2002)                   0   
3024     4043                At Close Range (1986)                   0   

      Action  Adventure  Animation  Children  Comedy  Crime  Documentary  \
264        0          0          0         0       1      0            0   
8903       0          0          0         0       0      0            0   
2713       1          1          0         0       0      0            0   
3817       0          0          0         0       0      0            0   
3024       0          0          0         0       0      1            0   

       ...     Film-Noir  Horror  IMAX  Musical  Mystery  Romance  Sci-Fi  \
264    ...             0       0     0        0        0        0       0   
8903   ...             0       0     0        0        0        0       0   
2713   ...             0       0     0        0        0        0       0   
3817   ...             0       0     0        0        0        0       0   
3024   ...             0       0     0        0        0        0       0   

      Thriller  War  Western  
264          0    0        0  
8903         0    0        0  
2713         1    0        0  
3817         0    0        0  
3024         0    0        0  

[5 rows x 22 columns]

movieId  (no genres listed)       Action    Adventure  \
count    9742.000000         9742.000000  9742.000000  9742.000000   
mean    42200.353623            0.003490     0.187641     0.129645   
std     52160.494854            0.058976     0.390445     0.335930   
min         1.000000            0.000000     0.000000     0.000000   
25%      3248.250000            0.000000     0.000000     0.000000   
50%      7300.000000            0.000000     0.000000     0.000000   
75%     76232.000000            0.000000     0.000000     0.000000   
max    193609.000000            1.000000     1.000000     1.000000   

         Animation     Children       Comedy        Crime  Documentary  \
count  9742.000000  9742.000000  9742.000000  9742.000000  9742.000000   
mean      0.062718     0.068158     0.385547     0.123075     0.045165   
std       0.242468     0.252031     0.486749     0.328541     0.207677   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     1.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

             Drama     ...         Film-Noir       Horror         IMAX  \
count  9742.000000     ...       9742.000000  9742.000000  9742.000000   
mean      0.447649     ...          0.008930     0.100390     0.016218   
std       0.497277     ...          0.094083     0.300535     0.126321   
min       0.000000     ...          0.000000     0.000000     0.000000   
25%       0.000000     ...          0.000000     0.000000     0.000000   
50%       0.000000     ...          0.000000     0.000000     0.000000   
75%       1.000000     ...          0.000000     0.000000     0.000000   
max       1.000000     ...          1.000000     1.000000     1.000000   

           Musical      Mystery      Romance       Sci-Fi     Thriller  \
count  9742.000000  9742.000000  9742.000000  9742.000000  9742.000000   
mean      0.034285     0.058817     0.163827     0.100595     0.194416   
std       0.181968     0.235295     0.370137     0.300808     0.395771   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     0.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

               War      Western  
count  9742.000000  9742.000000  
mean      0.039212     0.017142  
std       0.194108     0.129808  
min       0.000000     0.000000  
25%       0.000000     0.000000  
50%       0.000000     0.000000  
75%       0.000000     0.000000  
max       1.000000     1.000000  

[8 rows x 21 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9742 entries, 0 to 9741
Data columns (total 22 columns):
movieId               9742 non-null int64
title                 9742 non-null object
(no genres listed)    9742 non-null int64
Action                9742 non-null int64
Adventure             9742 non-null int64
Animation             9742 non-null int64
Children              9742 non-null int64
Comedy                9742 non-null int64
Crime                 9742 non-null int64
Documentary           9742 non-null int64
Drama                 9742 non-null int64
Fantasy               9742 non-null int64
Film-Noir             9742 non-null int64
Horror                9742 non-null int64
IMAX                  9742 non-null int64
Musical               9742 non-null int64
Mystery               9742 non-null int64
Romance               9742 non-null int64
Sci-Fi                9742 non-null int64
Thriller              9742 non-null int64
War                   9742 non-null int64
Western   

None

Analisando o cabeçalho, vemos que existe a coluna '(no genres listed)'. Vamos analisá-la um pouco:

In [6]:
print('Quantidade de filmes sem gênero: ' + str(movies['(no genres listed)'].sum()))
print('\n')
print(movies[movies['(no genres listed)'] == 1]['title'])

Quantidade de filmes sem gênero: 34


8517                                    La cravate (1957)
8684                                       Ben-hur (2016)
8687    Pirates of the Caribbean: Dead Men Tell No Tal...
8782                                    Superfast! (2015)
8836                                  Let It Be Me (1995)
8902                 Trevor Noah: African American (2013)
9033                                     Guardians (2016)
9053                                    Green Room (2015)
9070                       The Brand New Testament (2015)
9091                                           Hyena Road
9138    The Adventures of Sherlock Holmes and Doctor W...
9178                            A Cosmic Christmas (1977)
9217                                   Grease Live (2016)
9248                               Noin 7 veljestä (1968)
9259                                             Paterson
9307                          Ali Wong: Baby Cobra (2016)
9316                     A Midsumm

Para continuar a análise, vamos juntar os dois DataFrames:

In [7]:
ratings = pd.merge(ratings, movies, how='left', left_on ='movieId' ,right_on = 'movieId')
display(ratings.head())

userId  movieId  rating  timestamp                        title  \
0       1        1     4.0  964982703             Toy Story (1995)   
1       1        3     4.0  964981247      Grumpier Old Men (1995)   
2       1        6     4.0  964982224                  Heat (1995)   
3       1       47     5.0  964983815  Seven (a.k.a. Se7en) (1995)   
4       1       50     5.0  964982931   Usual Suspects, The (1995)   

   (no genres listed)  Action  Adventure  Animation  Children   ...     \
0                   0       0          1          1         1   ...      
1                   0       0          0          0         0   ...      
2                   0       1          0          0         0   ...      
3                   0       0          0          0         0   ...      
4                   0       0          0          0         0   ...      

   Film-Noir  Horror  IMAX  Musical  Mystery  Romance  Sci-Fi  Thriller  War  \
0          0       0     0        0        0        0       0         0    0   
1          0       0     0        0        0        1       0         0    0   
2          0       0     0        0        0        0       0         1    0   
3          0       0     0        0        1        0       0         1    0   
4          0       0     0        0        1        0       0         1    0   

   Western  
0        0  
1        0  
2        0  
3        0  
4        0  

[5 rows x 25 columns]

Agora vamos criar uma matriz esparsa de usuários x filmes com as notas como valores e depois converter em objeto Numpy para melhorar o processamento, visto que o Pandas foi construído usando Numpy

In [8]:
rp = ratings.pivot_table(columns='movieId', index='userId', values='rating')
rp = rp.fillna(0)
rp_mat = rp.values

rp_mat.shape

(610, 9724)

<h2>Matriz de Similaridade</h2>

No livro, é usada a função comentata abaixo, porém, existe uma forma muito mais eficiente utilizando cosine_similarity do Scikit. Por fim, vemos que ambas são no fundo, iguais.

In [9]:
#from scipy.spatial.distance import cosine

#m, n = rp.shape 
# User similarity matrix 
#mat_users = np.zeros((m, m)) 

#for i in range(m): 
#    for j in range(m): 
#        if i != j: 
#            mat_users[i][j] = (1- cosine(rp_mat[i,:], rp_mat[j,:])) 
#        else: 
#            mat_users[i][j] = 0. 

#pd_users = pd.DataFrame(mat_users, index =rp.index, columns= rp.index)

In [10]:
from sklearn.metrics.pairwise import cosine_similarity

pd_users = cosine_similarity(rp_mat)
np.fill_diagonal(pd_users,val=0)
pd_users = pd.DataFrame(pd_users, index =rp.index, columns = rp.index)

In [11]:
#display(pd_users.sample(n=5, random_state=123).iloc[:,:7])
display(pd_users.sample(n=5, random_state=123).iloc[:,:7])

userId         1         2         3         4         5         6         7
userId                                                                      
79      0.216426  0.010753  0.000000  0.120893  0.036173  0.043921  0.142875
12      0.016458  0.000000  0.000000  0.049945  0.058860  0.096056  0.045235
311     0.064405  0.000000  0.000000  0.059597  0.025718  0.025770  0.103642
283     0.127859  0.083982  0.000000  0.088068  0.149698  0.128061  0.129741
103     0.211480  0.144054  0.006157  0.142917  0.137857  0.117129  0.171629

In [12]:
# Encontrar usuários similares 
def top_similarusers(df, uid=16,n=5): 
    users = df.loc[uid, :].sort_values(ascending = False) 
    topn_users = users.iloc[:n, ] 
    topn_users = topn_users.rename('Score')     
    print ("Similar users as user:", uid) 
    print(pd.DataFrame(topn_users))
    print('\n')

top_similarusers(df=pd_users, uid=270, n=5)
top_similarusers(df=pd_users, uid=150, n=5)

#top_similarusers(df=pd_users, uid=17, n=5)

Similar users as user: 270
           Score
userId          
150     0.752312
206     0.752130
521     0.670239
389     0.665231
451     0.652090


Similar users as user: 150
           Score
userId          
270     0.752312
389     0.628609
521     0.585880
451     0.578992
206     0.574538




Que interessante, o usuário 270 é bem parecido com os usuários 150 e 206. Mas o usuário 150 não é muito parecido com o usuário 206!
<br/><br/>
Vamos encontrar agora os filmes mais bem votados por cada usuário:

In [13]:
#Encontrar melhores filmes do usuário
def topn_movieratings(df, uid, n_ratings=10):
    uid_ratings = df.loc[df['userId']==uid][['userId','rating','title']]
    uid_ratings = uid_ratings.sort_values(by='rating', ascending=False)
    print("Top",n_ratings,"filmes do usuário",uid)
    print(uid_ratings.iloc[:n_ratings,])
    print("\n")
    
topn_movieratings(df=ratings, uid=597, n_ratings=5)

Top 5 filmes do usuário 597
       userId  rating                                      title
92380     597     5.0                             Stripes (1981)
92356     597     5.0                 Grosse Pointe Blank (1997)
92333     597     5.0  Indiana Jones and the Last Crusade (1989)
92334     597     5.0                         Being There (1979)
92335     597     5.0                 Room with a View, A (1986)




Para fins didáticos, o livro usa a função abaixo para calcular a matriz de similaridade entre os filmes. Executando o código na minha máquina, levou 1 hora. Porém, a forma mais simples de fazer isso (e não ter que salvar um arquivo de 700mb em disco) é pegar a Transposta da matriz rp_mat e usar a função cosine_similarity novamente.

In [14]:
#from scipy.spatial.distance import cosine
#import time

#m, n = rp.shape 
#start_time = time.time()
    # Movie similarity matrix 
#mat_movies = np.zeros((n, n)) 

#for i in range(n): 
#    for j in range(n): 
#        if i != j: 
#            mat_movies[i][j] = (1- cosine(rp_mat[:,i], rp_mat[:,j])) 
#        else: 
#            mat_movies[i][j] = 0.

#print("--- %s seconds ---" % (time.time() - start_time))

In [15]:
#pd_movies = pd.DataFrame(mat_movies, index=rp.columns, columns= rp.columns)
#pd_movies.to_csv('pd_movies.csv', sep=',')
#pd_movies = pd.read_csv('pd_movies.csv', index_col='movieId')

In [16]:
rp_matTranspose = rp_mat.T
pd_movies = cosine_similarity(rp_matTranspose)
np.fill_diagonal(pd_movies,val=0)
pd_movies = pd.DataFrame(pd_movies, index = rp.columns, columns = rp.columns)

display(pd_movies.shape)

(9724, 9724)

In [17]:
display(pd_movies.sample(n=5, random_state=123).iloc[:,:7])
#display(pd_movies2.sample(n=5, random_state=123).iloc[:,:7])

movieId         1         2         3         4         5         6         7
movieId                                                                      
31692    0.000000  0.094208  0.000000  0.000000  0.000000  0.000000  0.000000
4814     0.098906  0.139707  0.065607  0.037883  0.063026  0.107146  0.066194
88272    0.060154  0.000000  0.000000  0.000000  0.000000  0.086888  0.000000
47721    0.051042  0.067292  0.060771  0.000000  0.000000  0.110590  0.102190
6347     0.000000  0.000000  0.000000  0.000000  0.000000  0.000000  0.000000

In [18]:
#Encontrar filmes similares
def topN_similarmovies(df, mID= 588, n=10):
    mID_ratings = df.iloc[mID,:].sort_values(ascending=False)
    topn_movies = pd.DataFrame(mID_ratings[:n,])
    topn_movies['index1'] = topn_movies.index
    topn_movies['index1'] = topn_movies['index1'].astype('int64')
    topn_movies = pd.merge(topn_movies, movies[['movieId', 'title']], how='left', left_on='index1', right_on='movieId')
    print('Filmes similares ao filme id:', mID, movies['title'][movies['movieId'] == mID].to_string(index=False))
    del topn_movies['index1']
    print(topn_movies)
    
topN_similarmovies(df=pd_movies, mID=589, n=15)

Filmes similares ao filme id: 589 Terminator 2: Judgment Day (1991)
         726  movieId                                              title
0   1.000000       55                                     Georgia (1995)
1   1.000000      406                                Federal Hill (1994)
2   1.000000      287                          Nina Takes a Lover (1994)
3   1.000000      722    Haunted World of Edward D. Wood Jr., The (1996)
4   1.000000      987                                       Bliss (1997)
5   1.000000      324                              Sum of Us, The (1994)
6   0.707107      331                                   Tom & Viv (1994)
7   0.707107      214              Before the Rain (Pred dozhdot) (1994)
8   0.707107      369          Mrs. Parker and the Vicious Circle (1994)
9   0.707107      321  Strawberry and Chocolate (Fresa y chocolate) (...
10  0.685994      113                            Before and After (1996)
11  0.606339      299                                   

Obviamente ocorreu algum erro e eu preciso rever o que aconteceu

<h2>Filtro Colaborativo usando ALS</h2>

In [19]:
A = rp.values
print("Shape of Original Sparse Matrix", A.shape)

Shape of Original Sparse Matrix (610, 9724)


Vamos agora criar uma matriz W com valor 0 ou 1 sendo que 1 sigfnifica que o usuário deu uma nota ao filme.

In [20]:
W = A > 0.5
W[W==True]=1
W[W==False]=0
W = W.astype(np.float64, copy=False)
print(W)

[[1. 0. 1. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [1. 1. 1. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]]


De forma similar, vamos criar uma matriz W_pred com os valores inversos para podermos gerar as sugestões de filmes (nós queremos só sugerir filmes que ele NÃO viu)

In [22]:
W_pred = A < 0.5
W_pred[W_pred == True] = 1
W_pred[W_pred == False] = 0
W_pred = W_pred.astype(np.float64, copy=False)
np.fill_diagonal(W_pred, val=0)